plugins/inputs/modbus/request.go

```diff
@@ -42,28 +189,85 @@ func groupFieldsToRequests(fields []field, tags map[string]string, maxBatchSize
-		needInterrupt = needInterrupt || f.length+current.length > maxBatchSize // too large
-
-		if !needInterrupt {
+		if current.length > 0 && f.address == current.address+current.length {
 			// Still safe to add the field to the current request
 			current.length += f.length
 			if !f.omit {
-				// Omit adding the field but use it for constructing the request.
 				current.fields = append(current.fields, f)
 			}
 			continue
 		}
 
 		// Finish the current request, add it to the list and construct a new one
-		requests = append(requests, current)
-		current = newRequest(f, tags)
+		if current.length > 0 {
+			groups = append(groups, current)
+		}
+		current = newRequest(f)
+	}
+	if current.length > 0 {
+		groups = append(groups, current)
+	}
+
+	var requests []request
+	switch optimization {
+	case "shrink":
+		// Shrink request by striping leading and trailing fields with an omit flag set
+		for _, g := range groups {
+			if len(g.fields) > 0 {
+				requests = append(requests, shrinkGroup(g, maxBatchSize)...)
+			}
+		}
+	case "rearrange":
+		// Allow rearranging fields between request in order to reduce the number of touched
+		// registers while keeping the number of requests
+		for _, g := range groups {
+			if len(g.fields) > 0 {
+				requests = append(requests, optimitzeGroup(g, maxBatchSize)...)
+			}
+		}
+	case "aggressive":
+		// Allow rearranging fields similar to "rearrange" but allow mixing of groups
+		// This might reduce the number of requests at the cost of more registers being touched.
+		var total request
+		for _, g := range groups {
+			if len(g.fields) > 0 {
+				total.fields = append(total.fields, g.fields...)
+			}
+		}
+		requests = optimitzeGroup(total, maxBatchSize)
+	case "max_insert":
+		// Similar to aggressive but keeps the number of touched registers bellow a threshold
+		requests = optimitzeGroupWithinLimits(fields, maxBatchSize, maxTouchedRegisters)
+	default:
+		// no optimization
+		for _, g := range groups {
+			if len(g.fields) > 0 {
+				requests = append(requests, splitMaxBatchSize(g, maxBatchSize)...)
+			}
+		}
+	}
+
+	// Copy the tags
+	for i := range requests {
+		requests[i].tags = make(map[string]string)
+		for k, v := range tags {
+			requests[i].tags[k] = v
+		}
+	}
+	fmt.Printf("[Modbus] Number of requests %d\n", len(requests))
+	totalTouchedFields := uint16(0)
+	for _, r := range requests {
+		// for idx, r := range requests {
+		totalTouchedFields += r.length / r.fields[0].length
+		// fmt.Printf("Request %d: address %d length %d for %d active fields: filling rate %v%%\n", idx, r.address, r.length, len(r.fields), int(100*float32(int(r.fields[0].length)*len(r.fields))/float32(r.length)))
+		// for yy, f := range r.fields {
+		// 	fmt.Printf("\t%d address %d\n", yy, f.address)
+		// }
 	}
-	requests = append(requests, current)
+	fmt.Printf("[Modbus] Total number of touched Registers %d\n", totalTouchedFields)
```

Comment:
No `Printf`ing in production code please! Remove this altogether!

Reply:
This is something I wanted to discuss with you.

Would it be acceptable to have these `Printf` only when the flag `--test` is on? If so, how can I do that?

The reason I wish to make this information available to the user is that there is no way to know beforehand how many requests will be generated for a given value of `max_extra_registers` but there are many more values possible for `max_extra_registers` than the possible number of requests. For instance, if I take my configuration file with my 64 signals out of ~1200, depending on the value of `max_extra_register` I get 6 (same as aggressive), 7, 8, 9, 11, 19, 21 or 22 requests (same as for shrink or rearrange).
So in fact I do not have to test all 124 possible values of `max_extra_registers` but only a handful of them. But without reading the information in the `Printf`, I have no way to devise a useful strategy to find the best value for `max_extra_registers`.

In fact, I would like to suggest to add something in the readme explaining the strategy
1. use the `--test` flag to identify a list of `max_extra_registers` values leading to unique number of requests
2. set the acquisition `interval` to a very low value (e.g. 10ms)
3. acquire the data for 5min with each value of `max_extra_registers`
4. check the actually recorded data period, for instance in `influx` call `SELECT MEAN(dt) FROM (SELECT ELAPSED(<oneSignaName>,1ms) AS dt FROM <MeasurementName>`)
5. select the `max_extra_registers` that gives the lowest actually recorded data period 

Doing this, with my setup, I find that the optimum is with 7 requests, obtained with `max_extra_registers` around 50.

Does this make sense to you?

Reply:
Any thoughts on this @srebhan ?

Reply:
The correct way is to pass in the `telegraf.Logger` facility of the modbus plugin and use `log.Debugf()` which produces the output if `--debug` is enabled. Please _DO NOT_ do any magic in here.

If you just want to print the number of requests, you can also print the number in `modbus.go`'s `Init()` function using `m.Log.Debugf` or if you want this more prominent using `m.Log.Infof`.